### ISSUE TITLE
How to resolve "First View: Test Error: Error configuring traffic-shaping" on docker linux.

### ISSUE THREAD

**Issue by JyotiPrakashMallick** [WPO-Foundation/wptagent#89]
Have installed VMware on top I am running ubuntu and two docker I am running one is agent and a server. Whenever I am trying to run a test I am getting "First View: Test Error: Error configuring traffic-shaping".

I have pulled the following two images as server and agent.
https://hub.docker.com/r/webpagetest/server/
https://hub.docker.com/r/webpagetest/agent/

Please refer the screenshot attached.
![capture](https://user-images.githubusercontent.com/6301012/37405898-e03a8ba0-27bb-11e8-968d-db88801d9834.PNG)
![capture1](https://user-images.githubusercontent.com/6301012/37405899-e07d91d4-27bb-11e8-8258-7d3d6b238625.PNG)

**Comment by JyotiPrakashMallick** [comment]
Log file showing me:
2018/03/15 16:04:09 - Test Created
2018/03/15 16:04:12 - Starting test (initiated by tester ubuntu-172.17.0.1)
2018/03/15 16:04:24 - Test Run Complete. Run: 1, Cached: 0, Done: 1, Tester: ubuntu-172.17.0.1, Test Run Error: "Error configuring traffic-shaping"
2018/03/15 16:04:24 - Extracting 4898 byte uploaded file '/tmp/phpUaeDXr' to './results/18/03/15/98/9'
2018/03/15 16:04:24 - Done Processing. Run: 1, Cached: 0, Done: 1, Tester: ubuntu-172.17.0.1, Test Run Error: "Error configuring traffic-shaping"
2018/03/15 16:04:24 - Test Complete

**Comment by JyotiPrakashMallick** [comment]
any thoughts on this ?

**Comment by JyotiPrakashMallick** [comment]
The above error resolved by running the following two command,
 $ sudo modprobe ifb numifbs=1 [For traffic shaping to work correctly, need to load the ifb module on the host]

After that I ran the agent 
docker run -d -p 4001:80 \
                --network="host" \
                -e "SERVER_URL=http://localhost:4000/work/" \
                -e "LOCATION=Test" \
                -e "SHAPER=none" \             [To disable traffic-shaping]
                --cap-add="NET_ADMIN" \ [To provide container NET_ADMIN capabilities]
                webpagetest/agent

**Comment by Ankit3794** [comment]
> The above error resolved by running the following two command,
> $ sudo modprobe ifb numifbs=1 [For traffic shaping to work correctly, need to load the ifb module on the host]
> 
> After that I ran the agent
> docker run -d -p 4001:80 
> --network="host" 
> -e "SERVER_URL=http://localhost:4000/work/" 
> -e "LOCATION=Test" 
> -e "SHAPER=none" \ [To disable traffic-shaping]
> --cap-add="NET_ADMIN" \ [To provide container NET_ADMIN capabilities]
> webpagetest/agent

Thanks @JyotiPrakashMallick this really helps

**Comment by firecodex** [comment]
Hi Jyoti, it looks like this solution worked for Ankit. But unfortunately, it didn't work for me.

Am I missing any points here?

My commands are:
$ docker run -d -p 4000:80 --rm webpagetest/server
$ modprobe ifb numifbs=1
$ docker run -d -p 4001:80 --network="host" -e "SERVER_URL=http://localhost:4000/work/" -e "LOCATION=Test" -e "SHAPER=none" --cap-add="NET_ADMIN" webpagetest/agent

**Comment by JyotiPrakashMallick** [comment]
@firecodex sorry to hear that this didn't work for you, could you please check the Webpage Installation Status Check and let me know here which of the status is missing or not running?
You can find the sample here "https://github.com/WPO-Foundation/wptagent/issues/289"

**Comment by tung-nt-niteco** [comment]
Hi,

We encountered this problem also, and we run all recommended commands above:
sudo modprobe ifb numifbs=1
and
sudo docker run --ipc=host -d --name wpt-agent00 -e "EXTRA_ARGS=--shaper none" -vvvv --network="host" -e "SERVER_URL=https://servergoeshere/work/" -e "LOCATION=locationgoeshere" -e KEY="keygoeshere" -e NAME="Speedcurve Agent Docker 00" --cap-add=NET_ADMIN webpagetest/agent

but we still got error "Test Error: Error configuring traffic-shaping" (and not only the first test failed).
Could you please advise?

**Comment by tung-nt-niteco** [comment]
Hi again, I see that I do not need -e "EXTRA_ARGS=--shaper none"

So the following commands work :)
sudo modprobe ifb numifbs=1
and
sudo docker run --ipc=host -d --name wpt-agent00 -vvvv --network="host" -e "SERVER_URL=https://servergoeshere/work/" -e "LOCATION=locationgoeshere" -e KEY="keygoeshere" -e NAME="Speedcurve Agent Docker 00" --cap-add=NET_ADMIN webpagetest/agent

**Comment by tung-nt-niteco** [comment]
It seems that I also had to disable traffic-shaping on WPT server (in locations.ini, set connectivity to the one that has no traffic-shaping defined in connectivity.ini)

**Comment by tung-nt-niteco** [comment]
After further investigation, we have found that if we separate docker agents to separate machines (1 agent/1 host), it will solve the traffic-shaping problem, and we do not have to disable this feature. Worked perfectly now.

**Comment by skmohammadi** [comment]
@tung-nt-niteco How did you do it?

**Comment by tung-nt-niteco** [comment]
> 
Do what, sorry? :)

**Comment by skmohammadi** [comment]
> After further investigation, we have found that if we separate docker agents to separate machines (1 agent/1 host), it will solve the traffic-shaping problem, and we do not have to disable this feature. Worked perfectly now.

@tung-nt-niteco as you said

**Comment by tung-nt-niteco** [comment]
@skmohammadi Yeah so we separated docker agents to separate machines (1 docker/1 VM) ^^

**Comment by skmohammadi** [comment]
@tung-nt-niteco In my case, I had run WPT-server on a remote ubuntu server, and an agent on localhost (WSL on windows).
but still got error "Test Error: Error configuring traffic-shaping".

**Comment by tung-nt-niteco** [comment]
@skmohammadi it is such a long time now so I don't remember everything, but I suggest:
- You contact SpeedCurve for support (I have to tell you that they might not be able to help right away but at least they have related experience).
- For a specific test, if you click through to the WebPageTest results and then click "View Test Log" to check the test log, it may tell you something.
- If you don't need traffic shaping, just disable it ^^  (of course, in that case, it will not simulate the network speed which is supposed to be different in mobile/desktop or LAN/3G/5G/etc. but you may just want to use the private agents because you want to test non-public websites for instance)
And so, in locations.ini of the WPT server, add:
**connectivity=Native** (to only use "No Traffic Shaping" connection)
or/and by adding -e "SHAPER=none" to the docker command of the agent:
sudo docker run --ipc=host -d --name wpt-agent --restart always --network="host" -e "SERVER_URL=http://your-wpt-server/work/" -e "LOCATION=your-location" **-e "SHAPER=none"** -e KEY="keyGoesHere" --cap-add=NET_ADMIN webpagetest/agent

Sorry but that is all knowledge I have. If you resolve this problem later on, please share your experience.

**Comment by skmohammadi** [comment]
After a lot of struggle, I found that:

- Traffic-Shaping relies on the IFB network interface, so the agent server network should support it.
- In the case of WSL2, IFB doesn't seem to be available.
- It may be necessary for Docker containers to run Agent server as follows (notice to use of **--cap-add=ALL**)

```
docker run -d --cap-add=ALL -p 4001:80 --network="host" -e "SERVER_URL=http://localhost:4000/work/" -e "LOCATION=Test" webpagetest/agent
```
It worked like a charm for me.
